main/src/ca/uwaterloo/flix/language/phase/Stratifier.scala

```diff
@@ -747,17 +739,38 @@ object Stratifier extends Phase[Root, Root] {
   /**
     * Restricts the given constraint graph `dg` to the predicates that occur in the given type `tpe`.
     */
-  private def restrict(dg: ConstraintGraph, tpe: Type): ConstraintGraph = {
-    val predSyms = predicateSymbolsOf(tpe)
-    dg.restrict(predSyms)
+  private def restrict(dg: ConstraintGraph, predSyms: Map[Name.Pred, Type]): ConstraintGraph = {
+    dg.restrict(predSyms, (t1, t2) => arityOf(t1) == arityOf(t2))
   }
 
   /**
-    * Returns the set of predicates that appears in the given row type `tpe`.
+    * Returns the map of predicates that appears in the given Schema `tpe`.
+    * A non-Schema type will result in an `InternalCompilerException`.
     */
-  private def predicateSymbolsOf(tpe: Type): Set[Name.Pred] = tpe.typeConstructors.foldLeft(Set.empty[Name.Pred]) {
-    case (acc, TypeConstructor.SchemaRowExtend(pred)) => acc + pred
-    case (acc, _) => acc
+  private def predicateSymbolsOf(tpe: Type): Map[Name.Pred, Type] = Type.eraseAliases(tpe) match {
+    case Type.Apply(Type.Cst(TypeConstructor.Schema, _), schemaRow, _) => predicateSymbolsOf(schemaRow, Map.empty)
+    case other => throw InternalCompilerException(s"Unexpected non-schema type $other")
   }
 
+  /**
+    * Recursively collects the predicates and their type from a nested `SchemaRowExtend` type.
+    */
+  @tailrec
+  private def predicateSymbolsOf(tpe: Type, acc: Map[Name.Pred, Type]): Map[Name.Pred, Type] = tpe match {
```

Comment:
Ok, I get it now. I would rename this to `visitType` and move it inside `predicateSymbolsOf`.